src/devices/machine/w5100.cpp

```diff
@@ -0,0 +1,4320 @@
+	snprintf(buffer, sizeof(buffer), "%d.%d.%d.%d",
+		(ip >> 24) & 0xff,
+		(ip >> 16) & 0xff,
+		(ip >> 8) & 0xff,
+		(ip >> 0) & 0xff
+	);
+	return buffer;
+}
+
+
+
+struct w5100_base_device::ip_info
+{
+	ip_info() = default;
+	ip_info(const ip_info &) = default;
+
+	void from_frame(const uint8_t *frame)
+	{
+		memcpy(dst_mac, frame + o_ETHERNET_SRC, 6);
+		memcpy(dst_ip, frame + 14 + o_IP_SRC_ADDRESS, 4);
+		proto = frame[14 + o_IP_PROTOCOL];
+	}
+
+	void from_socket(const uint8_t *socket)
+	{
+		switch (socket[Sn_MR] & 0x0f)
+		{
+			case Sn_MR_UDP:
+				proto = IP_PROTOCOL_UDP;
+				break;
+			case Sn_MR_TCP:
+				proto = IP_PROTOCOL_TCP;
+				break;
+			case Sn_MR_IPRAW:
+			default:
+				proto = socket[Sn_PROTO];
+		}
+		fragment = read16(socket + Sn_FRAGR0);
+		ttl = socket[Sn_TTL];
+		tos = socket[Sn_TOS];
+		memcpy(dst_ip, socket + Sn_DIPR0, 4);
+		memcpy(dst_mac, socket + Sn_DHAR0, 6);
+	}
+
+	ip_info &operator=(const ip_info &) = default;
+
+	uint16_t fragment = 0x4000;
+	uint8_t proto = 0;
+	uint8_t ttl = 128;
+	uint8_t tos = 0;
+	uint8_t dst_mac[6] = {};
+	uint8_t dst_ip[4] = {};
+};
+
+
+w5100_base_device::w5100_base_device(const machine_config &mconfig, device_type type, const char *tag, device_t *owner, uint32_t clock, dev_type device_type)
+	: device_t(mconfig, type, tag, owner, clock)
+	, device_network_interface(mconfig, *this, 10)
+	, device_memory_interface(mconfig, *this)
+	, m_device_type(device_type)
+	, m_irq_handler(*this)
+	, m_address_space_config("memory", ENDIANNESS_BIG, 8, 15, 0, address_map_constructor())
+{
+}
+
+void w5100_base_device::device_start()
+{
+	save_item(m_common_registers, "General Registers");
+	save_item(m_socket_registers[0], "Socket 0 Registers");
+	save_item(m_socket_registers[1], "Socket 1 Registers");
+	save_item(m_socket_registers[2], "Socket 2 Registers");
+	save_item(m_socket_registers[3], "Socket 3 Registers");
+	save_item(m_tx_buffer, "TX Buffer Memory");
+	save_item(m_rx_buffer, "RX Buffer Memory");
+
+	save_item(NAME(m_idm));
+	save_item(NAME(m_identification));
+	save_item(NAME(m_irq_state));
+
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_una), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_nxt), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, rcv_nxt), 4);
+
+	save_pointer(STRUCT_MEMBER(m_tcp, irs), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, iss), 4);
+
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_wl1), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_wl2), 4);
+
+	save_pointer(STRUCT_MEMBER(m_tcp, resend_seq), 4);
+
+	save_pointer(STRUCT_MEMBER(m_tcp, snd_wnd), 4);
+	save_pointer(STRUCT_MEMBER(m_tcp, rcv_wnd), 4);
+
+
+	m_interrupt_timer = timer_alloc(FUNC(w5100_base_device::interrupt_timer), this);
+	for (int sn = 0; sn < 4 ; ++sn)
+	{
+		emu_timer *t;
+		t = timer_alloc(FUNC(w5100_base_device::delayed_ack_timer), this);
+		t->set_param(sn);
+		m_delayed_ack_timers[sn] = t;
+
+		t = timer_alloc(FUNC(w5100_base_device::keep_alive_timer), this);
+		t->set_param(sn);
+		m_keep_alive_timers[sn] = t;
```

Comment:
What’s the purpose of the local `t`?  It just makes the code more error-prone because you’ve got more local mutable state.